### PR TITLE
Refactor have_seen_events to reduce OOMs

### PR DESCRIPTION
**Pull request by richvdh**
My server is currently OOMing in the middle of have_seen_events, so let's try to fix that.

### CODE REVIEW

File: synapse/storage/databases/main/events_worker.py [hunk 1]
```diff
@@ -1375,33 +1384,30 @@ async def _have_seen_events_dict(
         cache_results = {
             (rid, eid) for (rid, eid) in keys if self._get_event_cache.contains((eid,))
         }
-        results = {x: True for x in cache_results}
+        results = dict.fromkeys(cache_results, True)
```
Comment (richvdh) [inline synapse/storage/databases/main/events_worker.py:1387]: this was a driveby optimisation while I was in the area.

File: synapse/storage/databases/main/events_worker.py [hunk 2]
```diff
@@ -1356,14 +1356,23 @@ async def have_seen_events(
...
             The set of events we have already seen.
         """
-        res = await self._have_seen_events_dict(
-            (room_id, event_id) for event_id in event_ids
-        )
-        return {eid for ((_rid, eid), have_event) in res.items() if have_event}
+
+        # @cachedList chomps lots of memory if you call it with a big list, so
+        # we break it down. However, each batch requires its own index scan, so we make
+        # the batches as big as possible.
```
Comment (erikjohnston) [inline synapse/storage/databases/main/events_worker.py:1362]: OOI do we know *why* `@cachedList` is so heavyweight? That feels like it might be a useful thing to fix (separately)

**Review by erikjohnston — APPROVED** [review]

File: synapse/storage/databases/main/events_worker.py (hunk 2 above)
Comment (richvdh) [reply synapse/storage/databases/main/events_worker.py:1362]: I *think* it's because it creates a new Deferred for each (missing) entry in the list. Maybe?

I agree it would be a useful thing to investigate. I did consider trying to do add support for batching to `@cachedList`, but it looked scary.
